File: Programmers/P1844.js
function solution(maps) {
    var answer = 0;
    const dr = [-1,0,1,0];
    const dc = [0,1,0,-1];
    const visited = Array.from({length:maps.length},(v,i)=>Array(maps[0].length).fill(false));
    visited[0][0]=true;
    const queue = [[0,0]];
    let cnt =0;
    let find = false;
    while(queue.length!==0){
        cnt++;
        const size = queue.length;
        for(let i=0;i<size;i++){
            const now = queue.shift();
            for(let j=0;j<4;j++){
                if([now[0]+dr[j]]<0 || [now[0]+dr[j]]==maps.length ||[now[1]+dc[j]]<0 || [now[1]+dc[j]]==maps[0].length)continue;
                if(maps[now[0]+dr[j]][now[1]+dc[j]]==0)continue;
                if(visited[now[0]+dr[j]][now[1]+dc[j]]==false){
                    if([now[0]+dr[j]]==maps.length-1 && [now[1]+dc[j]]==maps[0].length-1){
                       find = true;
                        cnt++;
                        return cnt;
                    }
                    visited[now[0]+dr[j]][now[1]+dc[j]]=true;
                    queue.push([(now[0]+dr[j]),(now[1]+dc[j])]); 
                }
            }
        }
        
    }
    return -1;
}

File: Programmers/P1844.java
import java.util.*;
class Solution {
    public static int[] dr ={-1,0,1,0};
    public static int[] dc ={0,1,0,-1};
    
    public int solution(int[][] maps) {
        int answer = 0;
        
        Queue<int[]> queue = new LinkedList<int[]>();
        boolean[][] visited = new boolean[maps.length][maps[0].length];
        
        queue.add(new int[] {0,0});
        visited[0][0]=true;
        boolean find = false;
        
        while(!queue.isEmpty()){
            int size = queue.size();
            answer++;
            for(int i=0;i<size;i++){
                int[] now = queue.poll();
                if(now[0]==maps.length-1 && now[1]==maps[0].length-1){
                    find = true;
                    break;
                }
                for(int j=0;j<4;j++){
                    if(now[0]+dr[j]<0 || now[0]+dr[j]==maps.length||now[1]+dc[j]<0 || now[1]+dc[j]==maps[0].length)continue;
                    if(maps[now[0]+dr[j]][now[1]+dc[j]]==0)continue;
                    if(visited[now[0]+dr[j]][now[1]+dc[j]])continue;
                    
                    visited[now[0]+dr[j]][now[1]+dc[j]]=true;
                    queue.add(new int[] {now[0]+dr[j],now[1]+dc[j]});
                }
                if(find)break;
            }
            if(find)break;
        }
        return find?answer:-1;
    }
}
